In [2]:
import pandas as pd

# Cargamos datos y primer vistazo (ajustar la ruta según tus directorios)
df_main = pd.read_csv(r"C:\Users\Yago\Documents\Basketmática\ProyectoKing_Junio2024\lebron_shot_data.csv")
pd.set_option('display.max.columns', 25)

df_main.tail()

GAME_ID  GAME_EVENT_ID  PLAYER_ID   PLAYER_NAME     TEAM_ID  \
29306  22301230            233       2544  LeBron James  1610612747   
29307  22301230            300       2544  LeBron James  1610612747   
29308  22301230            340       2544  LeBron James  1610612747   
29309  22301230            380       2544  LeBron James  1610612747   
29310  22301230            410       2544  LeBron James  1610612747   

                TEAM_NAME  PERIOD  MINUTES_REMAINING  SECONDS_REMAINING  \
29306  Los Angeles Lakers       2                  7                  6   
29307  Los Angeles Lakers       2                  2                 30   
29308  Los Angeles Lakers       3                 11                 24   
29309  Los Angeles Lakers       3                  8                  5   
29310  Los Angeles Lakers       3                  6                 19   

        EVENT_TYPE                     ACTION_TYPE       SHOT_TYPE  \
29306    Made Shot      Driving Floating Jump Shot  2PT Field Goal   
29307    Made Shot  Running Finger Roll Layup Shot  2PT Field Goal   
29308    Made Shot  Driving Finger Roll Layup Shot  2PT Field Goal   
29309    Made Shot                Pullup Jump shot  3PT Field Goal   
29310  Missed Shot      Running Reverse Layup Shot  2PT Field Goal   

             SHOT_ZONE_BASIC SHOT_ZONE_AREA  SHOT_ZONE_RANGE  SHOT_DISTANCE  \
29306  In The Paint (Non-RA)      Center(C)         8-16 ft.              9   
29307        Restricted Area      Center(C)  Less Than 8 ft.              1   
29308        Restricted Area      Center(C)  Less Than 8 ft.              3   
29309      Above the Break 3      Center(C)          24+ ft.             26   
29310        Restricted Area      Center(C)  Less Than 8 ft.              1   

       LOC_X  LOC_Y  SHOT_ATTEMPTED_FLAG  SHOT_MADE_FLAG  GAME_DATE  HTM  VTM  \
29306     -1     91                    1               1   20231207  LAL  NOP   
29307     10      2                    1               1   20231207  LAL  NOP   
29308    -36      1                    1               1   20231207  LAL  NOP   
29309     23    269                    1               1   20231207  LAL  NOP   
29310    -12     -4                    1               0   20231207  LAL  NOP   

        SEASON  
29306  2023-24  
29307  2023-24  
29308  2023-24  
29309  2023-24  
29310  2023-24

# Preprocesado de datos

In [3]:
# Partido en el que LeBron rompe el record (7 de Febrero de 2023)
df_main[df_main['GAME_DATE'] == 20230207]

GAME_ID  GAME_EVENT_ID  PLAYER_ID   PLAYER_NAME     TEAM_ID  \
27807  22200821              7       2544  LeBron James  1610612747   
27808  22200821             25       2544  LeBron James  1610612747   
27809  22200821             62       2544  LeBron James  1610612747   
27810  22200821             76       2544  LeBron James  1610612747   
27811  22200821             82       2544  LeBron James  1610612747   
27812  22200821             97       2544  LeBron James  1610612747   
27813  22200821            187       2544  LeBron James  1610612747   
27814  22200821            199       2544  LeBron James  1610612747   
27815  22200821            205       2544  LeBron James  1610612747   
27816  22200821            239       2544  LeBron James  1610612747   
27817  22200821            362       2544  LeBron James  1610612747   
27818  22200821            394       2544  LeBron James  1610612747   
27819  22200821            400       2544  LeBron James  1610612747   
27820  22200821            426       2544  LeBron James  1610612747   
27821  22200821            441       2544  LeBron James  1610612747   
27822  22200821            476       2544  LeBron James  1610612747   
27823  22200821            486       2544  LeBron James  1610612747   
27824  22200821            501       2544  LeBron James  1610612747   
27825  22200821            569       2544  LeBron James  1610612747   
27826  22200821            640       2544  LeBron James  1610612747   

                TEAM_NAME  PERIOD  MINUTES_REMAINING  SECONDS_REMAINING  \
27807  Los Angeles Lakers       1                 11                 43   
27808  Los Angeles Lakers       1                  9                 58   
27809  Los Angeles Lakers       1                  7                  3   
27810  Los Angeles Lakers       1                  6                 11   
27811  Los Angeles Lakers       1                  5                 40   
27812  Los Angeles Lakers       1                  4                 50   
27813  Los Angeles Lakers       2                 11                 39   
27814  Los Angeles Lakers       2                 10                 53   
27815  Los Angeles Lakers       2                 10                 19   
27816  Los Angeles Lakers       2                  7                 52   
27817  Los Angeles Lakers       3                 11                 26   
27818  Los Angeles Lakers       3                  8                 29   
27819  Los Angeles Lakers       3                  7                 56   
27820  Los Angeles Lakers       3                  5                 11   
27821  Los Angeles Lakers       3                  4                 33   
27822  Los Angeles Lakers       3                  2                  0   
27823  Los Angeles Lakers       3                  1                  6   
27824  Los Angeles Lakers       3                  0                 10   
27825  Los Angeles Lakers       4                  7                 26   
27826  Los Angeles Lakers       4                  1                 51   

        EVENT_TYPE                 ACTION_TYPE       SHOT_TYPE  \
27807  Missed Shot  Driving Floating Jump Shot  2PT Field Goal   
27808  Missed Shot         Step Back Jump shot  2PT Field Goal   
27809    Made Shot         Step Back Jump shot  3PT Field Goal   
27810  Missed Shot                   Jump Shot  3PT Field Goal   
27811    Made Shot          Cutting Layup Shot  2PT Field Goal   
27812    Made Shot          Driving Layup Shot  2PT Field Goal   
27813  Missed Shot          Driving Layup Shot  2PT Field Goal   
27814    Made Shot          Running Layup Shot  2PT Field Goal   
27815    Made Shot          Driving Layup Shot  2PT Field Goal   
27816    Made Shot            Pullup Jump shot  3PT Field Goal   
27817  Missed Shot                   Jump Shot  2PT Field Goal   
27818    Made Shot                   Jump Shot  3PT Field Goal   
27819    Made Shot   Running Pull-Up Jump Shot  3PT Field Goal   
27820    Made Sh

Vemos que corresponde a la fila 27824 (o su tiro 27824), ya que su tiro fue a falta de 10 segundos para terminar el tercer cuarto. El análisis se hará desde el inicio de su carrera hasta este tiro

In [4]:
# Nuevo dataframe con los tiros hasta el tiro con el que rompe el récord
df_main = df_main[:-1486]
df_main.tail()

GAME_ID  GAME_EVENT_ID  PLAYER_ID   PLAYER_NAME     TEAM_ID  \
27820  22200821            426       2544  LeBron James  1610612747   
27821  22200821            441       2544  LeBron James  1610612747   
27822  22200821            476       2544  LeBron James  1610612747   
27823  22200821            486       2544  LeBron James  1610612747   
27824  22200821            501       2544  LeBron James  1610612747   

                TEAM_NAME  PERIOD  MINUTES_REMAINING  SECONDS_REMAINING  \
27820  Los Angeles Lakers       3                  5                 11   
27821  Los Angeles Lakers       3                  4                 33   
27822  Los Angeles Lakers       3                  2                  0   
27823  Los Angeles Lakers       3                  1                  6   
27824  Los Angeles Lakers       3                  0                 10   

        EVENT_TYPE               ACTION_TYPE       SHOT_TYPE  \
27820    Made Shot        Running Layup Shot  2PT Field Goal   
27821  Missed Shot        Fadeaway Bank shot  2PT Field Goal   
27822    Made Shot      Alley Oop Layup shot  2PT Field Goal   
27823    Made Shot        Running Layup Shot  2PT Field Goal   
27824    Made Shot  Turnaround Fadeaway shot  2PT Field Goal   

             SHOT_ZONE_BASIC SHOT_ZONE_AREA  SHOT_ZONE_RANGE  SHOT_DISTANCE  \
27820        Restricted Area      Center(C)  Less Than 8 ft.              3   
27821        Restricted Area      Center(C)  Less Than 8 ft.              2   
27822        Restricted Area      Center(C)  Less Than 8 ft.              2   
27823        Restricted Area      Center(C)  Less Than 8 ft.              2   
27824  In The Paint (Non-RA)      Center(C)         8-16 ft.             13   

       LOC_X  LOC_Y  SHOT_ATTEMPTED_FLAG  SHOT_MADE_FLAG  GAME_DATE  HTM  VTM  \
27820     18     26                    1               1   20230207  LAL  OKC   
27821     21      7                    1               0   20230207  LAL  OKC   
27822     12     25                    1               1   20230207  LAL  OKC   
27823     -2     28                    1               1   20230207  LAL  OKC   
27824    -48    129                    1               1   20230207  LAL  OKC   

        SEASON  
27820  2023-24  
27821  2023-24  
27822  2023-24  
27823  2023-24  
27824  2023-24

In [5]:
# Eliminamos columnas poco relevantes para el análisis
df_main.drop(columns = ['GAME_ID', 'GAME_EVENT_ID', 'PLAYER_ID', 'PLAYER_NAME', 'TEAM_ID', 'PERIOD', 'MINUTES_REMAINING', 'SECONDS_REMAINING', 'EVENT_TYPE', 'SHOT_ZONE_RANGE', 'SHOT_ATTEMPTED_FLAG', 'GAME_DATE', 'HTM', 'SEASON'], inplace=True)
df_main.head()

TEAM_NAME         ACTION_TYPE       SHOT_TYPE  \
0  Cleveland Cavaliers           Jump Shot  2PT Field Goal   
1  Cleveland Cavaliers           Jump Shot  2PT Field Goal   
2  Cleveland Cavaliers  Fadeaway Jump Shot  2PT Field Goal   
3  Cleveland Cavaliers           Jump Shot  2PT Field Goal   
4  Cleveland Cavaliers   Running Jump Shot  2PT Field Goal   

         SHOT_ZONE_BASIC SHOT_ZONE_AREA  SHOT_DISTANCE  LOC_X  LOC_Y  \
0              Mid-Range  Right Side(R)             15    158      0   
1              Mid-Range   Left Side(L)             13   -131      2   
2              Mid-Range  Right Side(R)             16    163      5   
3              Mid-Range   Left Side(L)             14   -118     92   
4  In The Paint (Non-RA)      Center(C)              5     54     22   

   SHOT_MADE_FLAG  VTM  
0               1  CLE  
1               1  CLE  
2               1  CLE  
3               0  CLE  
4               1  CLE

# 1. Métricas clave

In [6]:
# Función para calcular porcentaje de tiros anotados
def calcular_porcentaje_tiros(anotados, intentados):
    return (anotados / intentados) * 100 if intentados != 0 else 0

# Tiros de campo (T2 + T3)
tiros_totales_intentados = len(df_main)
tiros_totales_anotados = len(df_main[df_main['SHOT_MADE_FLAG'] == 1])
porcentaje_tiros_campo = calcular_porcentaje_tiros(tiros_totales_anotados, tiros_totales_intentados)
print(f"%TC: {porcentaje_tiros_campo:.2f}")

# Tiros de 2 y tiros de 3
df_tiros_dos = df_main[df_main['SHOT_TYPE'] == '2PT Field Goal']
df_tiros_tres = df_main[df_main['SHOT_TYPE'] == '3PT Field Goal']

tiros_dos_intentados = len(df_tiros_dos)
tiros_tres_intentados = len(df_tiros_tres)
tiros_dos_anotados = len(df_tiros_dos[df_tiros_dos['SHOT_MADE_FLAG'] == 1])
tiros_tres_anotados = len(df_tiros_tres[df_tiros_tres['SHOT_MADE_FLAG'] == 1])

porcentaje_tiros_dos = calcular_porcentaje_tiros(tiros_dos_anotados, tiros_dos_intentados)
porcentaje_tiros_tres = calcular_porcentaje_tiros(tiros_tres_anotados, tiros_tres_intentados)

print(f"%T2: {porcentaje_tiros_dos:.2f}")
print(f"%T3: {porcentaje_tiros_tres:.2f}")

# Puntos totales y ratio de tiros libres
puntos_totales = 38389
puntos_TC = 2 * tiros_dos_anotados + 3 * tiros_tres_anotados
puntos_TL = puntos_totales - puntos_TC
ratio_TL = (puntos_TL / puntos_totales) * 100

print(f"Ratio TL: {ratio_TL:.2f}")

%TC: 50.50
%T2: 55.38
%T3: 34.46
Ratio TL: 20.96


In [7]:
# Métricas avanzadas

# Effective Field Goal Percentage
EFG = (int(tiros_dos_anotados) + 1.5*int(tiros_tres_anotados))/int(tiros_totales_intentados)
print("%eFG: ", 100*EFG)

# 2 Point Rate
twoPR = (int(tiros_dos_intentados))/int(tiros_totales_intentados)
print("2PR: ", twoPR*100)

# 3 Point Rate
threePR = (int(tiros_tres_intentados))/int(tiros_totales_intentados)
print("3PR: ", threePR*100)

# Points Per Attempt
puntos = 2*int(tiros_dos_anotados) + 3*int(tiros_tres_anotados)
PPA = puntos/int(tiros_totales_intentados)
print("PPA: ", PPA)

%eFG:  54.521114106019766
2PR:  76.67205750224618
3PR:  23.327942497753817
PPA:  1.0904222821203953


# 2. Distribuciones

In [8]:
# Función para calcular porcentajes de tiro (intentados y anotados) según la zona
def calcular_porcentaje_tiros_por_zona(df, zona, total_tiros):
    tiros_zona = df[df['SHOT_ZONE_BASIC'] == zona]
    porcentaje_intentados = (len(tiros_zona) / total_tiros) * 100
    tiros_anotados = len(tiros_zona[tiros_zona['SHOT_MADE_FLAG'] == 1])
    porcentaje_anotados = (tiros_anotados / len(tiros_zona) * 100) if len(tiros_zona) > 0 else 0
    return porcentaje_intentados, porcentaje_anotados

# Total de tiros intentados
tiros_totales_intentados = len(df_main)

# Definir zonas de tiro
zonas_tiro = ['Mid-Range', 'In The Paint (Non-RA)', 'Restricted Area', 'Backcourt', 
              'Above the Break 3', 'Left Corner 3', 'Right Corner 3']

# Calcular y mostrar porcentajes de tiros por zona
for zona in zonas_tiro:
    porcentaje_intentados, porcentaje_anotados = calcular_porcentaje_tiros_por_zona(df_main, zona, tiros_totales_intentados)
    print(f"{zona}: {porcentaje_intentados:.2f}% intentados, {porcentaje_anotados:.2f}% anotados")

Mid-Range: 27.49% intentados, 38.16% anotados
In The Paint (Non-RA): 10.73% intentados, 40.04% anotados
Restricted Area: 38.45% intentados, 71.98% anotados
Backcourt: 0.12% intentados, 6.06% anotados
Above the Break 3: 20.22% intentados, 34.13% anotados
Left Corner 3: 1.43% intentados, 39.70% anotados
Right Corner 3: 1.56% intentados, 36.03% anotados


In [14]:
# Guardamos el DataFrame actualizado en un archivo Excel para luego usarlo en Tableau (ajustar la ruta según tus directorios)
df_main.to_excel(r"C:\Users\Yago\Documents\Basketmática\lebron_shot_data_cleaned.xlsx")